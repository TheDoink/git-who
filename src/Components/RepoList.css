#repoList {
	overflow-y: auto;
}

#noReposMessage {
	text-align: center;
}

#getMore {
	text-align: center;
	color: #078800;
}

#getMore #getMoreIcon {
	font-size: 10em;
}

#getMore #getMoreText {
	font-size:3em;
}