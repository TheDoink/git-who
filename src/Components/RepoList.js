import React, { Component } from 'react';

import RepoItem from './RepoItem';
import './RepoList.css';

import $ from 'jquery';

class RepoList extends Component {

  constructor(props) {
    super(props);

    this.state = {
      repoList: [], // This list of commits for this repo
      pageNum: 1,     // The page number we're on (30 items per page)
      loading: false  // Whether or not we're in the process of getting more
    };
  };

  getRepos(gitUser, pageNum) {

    // Ensure we actually have a good user
    if(gitUser.login) {
      this.setState({loading: true}, function() {
        // Try and get that username...
        $.get(`https://api.github.com/users/${gitUser.login}/repos?visibility=public&sort=updated&direction=desc&page=${pageNum}`,

          // If we get a good response...
          (success, text, obj) => {

            this.state.repoList = this.state.repoList.concat(success);

            // Append the new repos to the old repos, and update the pageNum
            this.setState({repoList: this.state.repoList, pageNum: pageNum, loading:false}, function() {
              console.log(this.state);
            });

          })

          // If we get anything else...
          .fail((failure, text, error) => {
            
          });
      });
    }
  }

  /* Only get repos if we have a new username, and if it's good */
  componentWillReceiveProps(nextProps) {
    if(nextProps.gitUser) {

      this.getRepos(nextProps.gitUser, 1);
    }

    // Even if no good username, reset the list
    this.setState({repoList: [], pageNum: 1}, function() {});
  };

  render() {
    return (
      <div id="repoList">
        {this.state.repoList.map(item => (
          <RepoItem key={item.url} repo={item}>
            
          </RepoItem>
          
        ))}

        {/* A button that is used to get additional repos, only shown when there is a valid user, and when not loading */}
        <div id="getMore" onClick={() => this.getRepos(this.props.gitUser, this.state.pageNum+1)} className={this.props.gitUser && this.state.repoList.length < this.props.gitUser.public_repos && !this.state.loading ? '' : 'hidden'}>
          <i id="getMoreIcon"  className="fa fa-arrow-circle-o-down"></i>
          <div id="getMoreText">Get More?</div>
        </div>
      </div>
    );
  }
}

RepoList.propTypes = {
  gitUser: React.PropTypes.object
}

export default RepoList;
