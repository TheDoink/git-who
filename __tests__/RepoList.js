// var NameLink = require('../src/Components/NameLink');
import React from 'react';
import renderer from 'react-test-renderer';
import RepoList from '../src/Components/RepoList';

const mockUser = {
  "login": "hank",
  "id": 8534,
  "avatar_url": "https://avatars1.githubusercontent.com/u/8534?v=4",
  "gravatar_id": "",
  "url": "https://api.github.com/users/hank",
  "html_url": "https://github.com/hank",
  "followers_url": "https://api.github.com/users/hank/followers",
  "following_url": "https://api.github.com/users/hank/following{/other_user}",
  "gists_url": "https://api.github.com/users/hank/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/hank/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/hank/subscriptions",
  "organizations_url": "https://api.github.com/users/hank/orgs",
  "repos_url": "https://api.github.com/users/hank/repos",
  "events_url": "https://api.github.com/users/hank/events{/privacy}",
  "received_events_url": "https://api.github.com/users/hank/received_events",
  "type": "User",
  "site_admin": false
};

it('lines up with the snapshot', () => {

  // Testing visible gists & repos
  let testMe = renderer.create(<RepoList gitUser={mockUser} repoOrGist="repo" shouldHide="false"></RepoList>).toJSON();
  expect(testMe).toMatchSnapshot();

  testMe = renderer.create(<RepoList gitUser={mockUser} repoOrGist="gist" shouldHide="false"></RepoList>).toJSON();
  expect(testMe).toMatchSnapshot();

  // Testing hidden gists & repos
  testMe = renderer.create(<RepoList gitUser={mockUser} repoOrGist="repo" shouldHide="true"></RepoList>).toJSON();
  expect(testMe).toMatchSnapshot();

  testMe = renderer.create(<RepoList gitUser={mockUser} repoOrGist="gist" shouldHide="true"></RepoList>).toJSON();
  expect(testMe).toMatchSnapshot();
});
