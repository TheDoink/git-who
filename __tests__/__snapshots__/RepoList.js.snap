// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lines up with the snapshot 1`] = `
<div>
  <div
    className={false}
  >
    <div
      className="repoList "
    >
      <div
        className="noItems "
      >
        This user has no 
        repo
        s yet!
      </div>
      <div
        className="getMore "
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-circle-o-down getMoreIcon"
        />
        <div
          className="getMoreText"
        >
          Get More?
        </div>
      </div>
    </div>
  </div>
  <div
    className="hidden"
  >
    <div
      className="failText"
    >
      A problem occured!
    </div>
  </div>
</div>
`;

exports[`lines up with the snapshot 2`] = `
<div>
  <div
    className={false}
  >
    <div
      className="repoList "
    >
      <div
        className="noItems "
      >
        This user has no 
        gist
        s yet!
      </div>
      <div
        className="getMore "
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-circle-o-down getMoreIcon"
        />
        <div
          className="getMoreText"
        >
          Get More?
        </div>
      </div>
    </div>
  </div>
  <div
    className="hidden"
  >
    <div
      className="failText"
    >
      A problem occured!
    </div>
  </div>
</div>
`;

exports[`lines up with the snapshot 3`] = `
<div>
  <div
    className={false}
  >
    <div
      className="repoList hidden"
    >
      <div
        className="noItems "
      >
        This user has no 
        repo
        s yet!
      </div>
      <div
        className="getMore "
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-circle-o-down getMoreIcon"
        />
        <div
          className="getMoreText"
        >
          Get More?
        </div>
      </div>
    </div>
  </div>
  <div
    className="hidden"
  >
    <div
      className="failText"
    >
      A problem occured!
    </div>
  </div>
</div>
`;

exports[`lines up with the snapshot 4`] = `
<div>
  <div
    className={false}
  >
    <div
      className="repoList hidden"
    >
      <div
        className="noItems "
      >
        This user has no 
        gist
        s yet!
      </div>
      <div
        className="getMore "
        onClick={[Function]}
      >
        <i
          className="fa fa-arrow-circle-o-down getMoreIcon"
        />
        <div
          className="getMoreText"
        >
          Get More?
        </div>
      </div>
    </div>
  </div>
  <div
    className="hidden"
  >
    <div
      className="failText"
    >
      A problem occured!
    </div>
  </div>
</div>
`;
